tests/integration/utils.py

```diff
@@ -463,3 +463,133 @@ def _validate_expected_dep_file_contents(dep_contents_file: Path, output_dir: Pa
+
+    >>> no_code_was_affected_by((Path('cachi2/core/config.py'),))
+    False
+    >>> no_code_was_affected_by((Path('tests/integration/test_gomod.py'),))
+    False
+    >>> no_code_was_affected_by((Path('README.md'),))
+    True
+    >>> no_code_was_affected_by((Path('reqruiements.txt'),))
+    True
+    """
+    for changed_file in changes:
+        for possible_path in paths_to_code:
+            try:
+                changed_file.relative_to(possible_path)
+            except ValueError:
+                pass
+            else:  # Something was changed within paths_to_code.
+                return False
+    # None of the changed objects is in watched directories.
+    return True
+
+
+def extract_pm_name(path: Path) -> str:
+    return path.stem.lstrip("test_")
+
+
+def affects_pm(change: Path) -> bool:
+    """Check if a pm is affected.
+
+    >>> affects_pm(Path('cachi2/core/config.py'))
+    False
+    >>> affects_pm(Path('reqruiements.txt'))
+    False
+    >>> affects_pm(Path('tests/integration/test_gomod.py'))
+    True
+    >>> affects_pm(Path('cachi2/core/package_managers/rpm/main.py'))
+    True
+    >>> affects_pm(Path('cachi2/core/package_managers/general.py'))
+    True
+    """
+
+    def affects_pm_or_support_code(change: Path) -> bool:
+        return "package_managers" in change.parts
+
+    def affects_pm_tests(change: Path) -> bool:
+        return change.stem.lstrip("test_") in supported_pms
+
+    return affects_pm_or_support_code(change) or affects_pm_tests(change)
+
+
+def affected_package_managers(changes: tuple[Path, ...]) -> set[str]:
+    return set(extract_pm_name(c) for c in changes if affects_pm(c))
+
+
+def just_package_managers_were_affected_by(changes: tuple[Path, ...]) -> bool:
+    """Check that just package managers were affected.
+
+    >>> c = Path('cachi2/core/package_managers/pip.py'), Path('tests/integration/test_pip.py')
+    >>> just_package_managers_were_affected_by(c)
+    True
+    >>> c = Path('cachi2/core/package_managers/pip.py'), Path('tests/integration/utils.py')
+    >>> just_package_managers_were_affected_by(c)
+    False
+    >>> just_package_managers_were_affected_by((Path('cachi2/core/utils.py'),))
+    False
+    """
+    return all(affects_pm(c) for c in changes)
+
+
+# 'Any' was used to let sentinel objects be returned as well as a set of names without
+# introducing full-blown singleton classes for TEST_EVERYTHING and TEST_NOTHING.
+def determine_integration_tests_to_run_for(ref: str = "HEAD") -> Any:
+    """Check which tests to run basing on which files were changed in a commit."""
+    if is_a_release_candidate(ref):
+        return TEST_EVERYTHING
+    changes = retrieve_changed_files_from_git(ref)
+    if no_code_was_affected_by(changes):
+        return TEST_NOTHING
+    elif just_package_managers_were_affected_by(changes):
+        return affected_package_managers(changes)
+    else:
+        return TEST_EVERYTHING
```

Comment:
How does this fare with regular releases? What ensures all tests are run when a release is tagged? If I'm not missing anything then we have to list tags for the current ref and if a tag is set then we need run everything; which goes back to my earlier comment on detecting RCs...I don't think we need to do that specifically, we need to run everything on any tag.

Reply:
Oh, will if just fall through to the last `else` condition in that case? But then, when we tag a commit, that ref will only trigger a subset of tests based on files changed anyway, so I really do think that on tagging a release in git (which is currently admittedly not being done by the book IMO, but we know that) we should make sure we run **everything**.

Reply:
Basing on the current draft a release candidate note will trigger all tests (which we _maybe_ don't really need after all, see my comment above). L587 handles that. Once we are convinced that everything is OK we tag the RC and thus have it released.